### PR TITLE
stdio/printf: Fix some issues with printf

### PR DESCRIPTION
**Pull request by astalke**
## Description

1. Fix an issue with `format_parse` failing silently. After this change, it returns an error code.
2. Fix an issue with %f allocating too much memory.

## Motivation and Context

`printf` causes issues on some platforms with no MMU and small amount of memory.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: ia32-generic-qemu

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: stdio/bignum.c [hunk 1]
```diff
@@ -37,13 +38,28 @@ static void bignum_shiftrInternal(bignum_t *bn, const size_t shift)
...
+	else if (bn->size <= BIGNUM_STACK_BUFFER_SIZE) {
+		ptr = calloc(newSize, sizeof(*bn->data));
+		if (ptr != NULL) {
+			(void)memcpy(ptr, bn->data, bn->size * sizeof(*bn->data));
+			bn->size = newSize;
+			bn->data = ptr;
+			return 0;
+		}
+	}
+	else {
```
Comment (agkaminski) [inline stdio/bignum.c:54]: Case of `bn->size > newSize` and `bn->size == newSize` aren't handled

**Review by agkaminski — COMMENTED** [review]
One (imho small) remark about `bignum_resize` (I guess we don't really use it to shrink it's size, so it's generally ok, until someone uses it in some other place).

File: stdio/bignum.c (hunk 1 above)
Comment (gerard5) [reply stdio/bignum.c:54]: For me missing cases are:
* `bn->size == newSize`  -> no operation 
* `(newSize < bn->Size) && (bn->data != bn->stackBuffer)` -> shrink from `re/c/alloc` to `use-stack`.

I see that way, until `bn->size` and `newSize` are grater than `BIGNUM_STACK_BUFFER_SIZE`, the check `bn->size > newSize` (shrink) and `bn->size < newSize` (grow) are handled (catched) by this `else` block with `realloc()`, but `bn->size == newSize` case should be handled explicitly `just return 0`, of course `realloc()` with reassignement in this block will handle that but it's waste of cycles. However, the case of freeing `realloc'ed` memory when returning to the `use stack memory`case is definitely not handled: `newSize < bn->Size` when `bn->data != bn->stackBuffer`.

File: stdio/bignum.c [hunk 2]
```diff
@@ -104,21 +158,26 @@ static int bignum_cmpInternal(const bignum_t *first, const bignum_t *second)
 int bignum_init(bignum_t *bn, size_t size, bignum_data_t val)
 {
 	bignum_t result = { .data = NULL, .len = 1, .size = size };
-	result.data = calloc(result.size, sizeof(*result.data));
-	if (result.data != NULL) {
-		result.data[0] = val;
-		*bn = result;
-		return 0;
+	if (size > BIGNUM_STACK_BUFFER_SIZE) {
```
Comment (agkaminski) [inline stdio/bignum.c:161]: To simplify the logic, how about:
- init bn->data = bn->stackBuffer,
- init bn->size = BIGNUM_STACK_BUFFER_SIZE
- do bignum_resize(bn, size).

sidenote - `result` temp var seems redundant to me

**Review by agkaminski — APPROVED** [review]
LGTM

**Review by gerard5 — APPROVED** [review]
Verified on HW-NIL, looks good to me.

**Comment by gerard5** [comment]
@astalke @agkaminski After merging there is an issue with this commit https://github.com/phoenix-rtos/libphoenix/pull/249/files function should now return an `int` but returns no value (issue appeared probably due to automatic rebase during merge).

See https://github.com/phoenix-rtos/libphoenix/blob/master/stdio/format.c#L888

**Comment by astalke** [comment]
> @astalke @agkaminski After merging there is an issue with this commit https://github.com/phoenix-rtos/libphoenix/pull/249/files function should now return an `int` but returns no value (issue appeared probably due to automatic rebase during merge).
> 
> See https://github.com/phoenix-rtos/libphoenix/blob/master/stdio/format.c#L888

Fix here: https://github.com/phoenix-rtos/libphoenix/pull/252

**Pull request merged**